### PR TITLE
Make schedulers picklable by making lr_lambda fns global

### LINKED ISSUE #21689: Make schedulers picklable
### Feature request

Change lambda functions passed to `LambdaLR` in `get_constant_schedule`, `get_constant_schedule_with_warmup`, `get_linear_schedule_with_warmup`, `get_cosine_schedule_with_warmup`, `get_cosine_with_hard_restarts_schedule_with_warmup` and `get_polynomial_decay_schedule_with_warmup` to callable objects.

### Motivation

Python cannot serialize lambda and local functions. Torch created a workaround around this in their `state_dict` method of `LambdaLR` by not returning any non-picklable functions:
```python
        ...
        for idx, fn in enumerate(self.lr_lambdas):
            if not isinstance(fn, types.FunctionType):
                state_dict['lr_lambdas'][idx] = fn.__dict__.copy()

        return state_dict
```
While this approach is fine when LR schedule is constant and deterministic, it makes it impossible to change the schedule mid training dynamically using lambda functions since any changes will not be saved to checkpoints.

In my particular case I wanted to implement a dynamic LR schedule based on evaluation metrics. I've implemented a wrapper around `LambdaLR` that applies transformation `fn: float -> float` to existing LR schedule:
```python
class LambdaWrapper:

    def __init__(self, lr_lamda: Callable[[Union[float, int]], float], wrapper_function: Callable[[float], float]):
        self._wrapper_function = wrapper_function
        self._lr_lambda = lr_lamda

    def __call__(self, x: Union[float, int]):
        return self._wrapper_function(self._lr_lambda(x))


class DynamicScheduler:

    def __init__(self, lr_scheduler: LambdaLR):
        self._scheduler = lr_scheduler

    def __getattr__(self, item):
        # Calling the super class to avoid recursion
        return getattr(super(DynamicScheduler, self).__getattribute__('_scheduler'), item)

    def wrap_schedule(self, fn: Callable[[float], float]):
        """If you want this object to be picklable, pass only picklable callable objects as `fn`!"""
        wrappers_builder = partial(LambdaWrapper, wrapper_function=fn)  # wrap in callable object to preserve picklability
        self._scheduler.lr_lambdas = list(map(wrappers_builder, self._scheduler.lr_lambdas))
```
I've taken special care to preserve picklability, however, since `LambdaLR` instances created by `transformers` library hold lambda and local functions in them, pickling of `DynamicScheduler` (as well as it's state, which is the same as the wrapped `LambdaLR` state) fails.

While reimplementing dynamic scheduling with lambda functions will allow the `torch` workaround that handles lambda functions in scheduler, the whole point of dynamic scheduling will be lost since the complex dynamically constructed lambdas: `f_n(f_n-1(...f_1(schedule(x))...))` will fall back to their default state: `schedule(x)`.

Here is the callback I use to track evaluation metrics for anyone interested:
```python
def get_warmup_steps(args: TrainingArguments, state: TrainerState) -> int:
    return (
        args.warmup_steps
        if args.warmup_steps > 0
        else math.ceil(state.max_steps * args.warmup_ratio)
    )


class DecreaseLRTransformer:
    def __init__(self, decrease_ratio: float):
        if decrease_ratio < 0.0 or decrease_ratio > 1.0:
            raise ValueError('Decrease ratio should be within [1.0, 0.0]')
        self._decrease_ratio = decrease_ratio

    def __call__(self, lr: float):
        return self._decrease_ratio * lr


# Developer notice (may change in the future versions of transformers):
# All kwargs have the following fields set: model, tokenizer, optimizer, lr_scheduler, train_dataloader, eval_dataloader

class LRDecreaseCallback(TrainerCallback):
    """
    A [`TrainerCallback`] that handles learning rate decrease based on evaluation metrics.
    """

    def __init__(self, decrease_ratio: float, patience: int, *, decrease_on_warmup: bool = False, decrease_threshold: float = 0.0):
        self._transformer = DecreaseLRTransformer(decrease_ratio)
        self._patience = patience
        self._decrease_on_warmup = decrease_on_warmup
        self._decrease_threshold = decrease_threshold

        self._failed_checks = 0

    def _metric_improved(self, new_metric: float, old_metric: float, *, greater_is_better: bool = True) -> bool:
        operator = np.greater if greater_is_better else np.less
        return operator(new_metric, old_metric) and abs(new_metric - old_metric) > self._decrease_threshold

    def check_metric_value(self, args: TrainingArguments, state: TrainerState, control: TrainerControl, metric_value: float):
        # best_metric is set by code for load_best_model
        no_metric = (state.best_metric is None)

        warmup_steps = get_warmup_steps(args, state)
        skip_warmup = (self._decrease_on_warmup and warmup_steps >= state.global_step)

        if skip_warmup:
            return

        if no_metric or self._metric_improved(metric_value, state.best_metric, greater_is_better=args.greater_is_better):
            self._failed_checks = 0
            control.should_save = True
        else:
            self._failed_checks += 1

    def on_train_begin(self, args: TrainingArguments, state: TrainerState, control: TrainerControl, **kwargs):
        if args.metric_for_best_model is None:
            raise ValueError(f"{self.__class__.__name__} requires metric_for_best_model to be defined defined")

        if args.evaluation_strategy == IntervalStrategy.NO:
            raise ValueError(f"{self.__class__.__name__} requires IntervalStrategy of steps or epoch")

    def on_evaluate(self, args: TrainingArguments, state: TrainerState, control: TrainerControl, **kwargs):
        metrics: Dict[str, float] = kwargs['metrics']
        lr_scheduler = kwargs['lr_scheduler']
        if not isinstance(lr_scheduler, DynamicScheduler):
            logger.warning(f'{self.__class__.__name__} is not compatible with {lr_scheduler.__class__.__name__} scheduler! '
                           f'Wrap your scheduler with {DynamicScheduler.__class__.__name__} to change LR dynamically. '
                           f'{self.__class__.__name__} is disabled!')
            return

        metric_to_check = args.metric_for_best_model
        if not metric_to_check.startswith("eval_"):
            metric_to_check = f"eval_{metric_to_check}"
        metric_value = metrics.get(metric_to_check)

        if metric_value is None:
            logger.warning(f"{self.__class__.__name__} required metric_for_best_model, "
                           f"but did not find {metric_to_check} in evaluation metrics. {self.__class__.__name__} is disabled!")
            return

        self.check_metric_value(args, state, control, metric_value)
        if self._failed_checks >= self._patience:
            lr_scheduler.wrap_schedule(self._transformer)
            self._failed_checks = 0

    def on_log(self, args: TrainingArguments, state: TrainerState, control: TrainerControl, **kwargs):
        logs: Dict[str, float] = kwargs['logs']
        logs['lr_decrease_patience'] = (self._patience - self._failed_checks) / self._patience
``` 

### Your contribution

The simplest and the cleanest workaround would be to make the local functions global:

Intead of:
```python
def get_linear_schedule_with_warmup(optimizer, num_warmup_steps, num_training_steps, last_epoch=-1):
    def lr_lambda(current_step: int):
        if current_step < num_warmup_steps:
            return float(current_step) / float(max(1, num_warmup_steps))
        return max(
            0.0, float(num_training_steps - current_step) / float(max(1, num_training_steps - num_warmup_steps))
        )

    return LambdaLR(optimizer, lr_lambda, last_epoch)
```

Do this:
```python
def _linear_schedule_with_warmup_step(current_step: int, *, num_warmup_steps: int, num_training_steps: int) -> float:
      if current_step < num_warmup_steps:
          return float(current_step) / float(max(1, num_warmup_steps))
      return max(
          0.0, float(num_training_steps - current_step) / float(max(1, num_training_steps - num_warmup_steps))
      )

def get_linear_schedule_with_warmup(optimizer, num_warmup_steps, num_training_steps, last_epoch=-1):
    schedule = partial(_linear_schedule_with_warmup_step, num_warmup_steps=num_warmup_steps, num_training_steps=num_training_steps)
    return LambdaLR(optimizer, schedule, last_epoch)
```
When created with global functions, partial function are picklable:
```python
>>>from functools import partial

>>>import pickle

>>>def f(x):
...    print(x)
    
>>>with open('f.pkl', 'wb') as file:
...    pickle.dump(partial(f, x='Dog'), file)
    
>>>with open('f.pkl', 'rb') as file:
...    unpickled_f = pickle.load(file)
    
>>>unpickled_f()
Dog
```

The fix is straightforward and I can create a PR. Nonetheless, it would be my first contribution so I might need some help along the way.

### PR DESCRIPTION
**Pull request by connor-henderson**
# What does this PR do?
Make schedulers picklable by making lr_lambda fns global, at the cost of the extra step of arg passing and using `partial`.

Closes #21689

Implements the change mentioned in the issue across the following functions in optimizations.py:
`get_constant_schedule`
`get_constant_schedule_with_warmup`
`get_linear_schedule_with_warmup`
`get_cosine_schedule_with_warmup`
`get_cosine_with_hard_restarts_schedule_with_warmup`
`get_inverse_sqrt_schedule`

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case. [link](https://github.com/huggingface/transformers/issues/21689)
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@sgugger

### CODE REVIEW

File: src/transformers/optimization.py [hunk 1]
```diff
@@ -44,7 +49,15 @@ def get_constant_schedule(optimizer: Optimizer, last_epoch: int = -1):
     Return:
         `torch.optim.lr_scheduler.LambdaLR` with the appropriate schedule.
     """
-    return LambdaLR(optimizer, lambda _: 1, last_epoch=last_epoch)
+
+    lr_lambda = partial(_get_constant_schedule_lr_lambda)
```
Comment (sgugger) [inline src/transformers/optimization.py:53]: Why use a partial here, there is no need?

**Review by sgugger — APPROVED** [review]
Thnaks for working on this!

**Comment by ViktorooReps** [comment]
It may be a good idea to test this feature:

In file `tests/optimization/test_optimization.py`:
Add class
```python
class LambdaScheduleWrapper:
    """See https://github.com/huggingface/transformers/issues/21689"""
    def __init__(self, fn):
        self.fn = fn

    def __call__(self, *args, **kwargs):
        return self.fn(*args, **kwargs)

    @classmethod
    def wrap_scheduler(cls, scheduler: LambdaLR):
        scheduler.lr_lambdas = list(map(cls, scheduler.lr_lambdas))
```

And wrap the schedulers before testing the reload process in `ScheduleInitTest.test_schedulers`:
```python
            ...
            scheduler = scheduler_func(self.optimizer, **kwargs)
++          LambdaScheduleWrapper.wrap_scheduler(scheduler)  # wrap to test picklability of the schedule
            lrs_2 = unwrap_and_save_reload_schedule(scheduler, self.num_steps)
            self.assertListEqual(lrs_1, lrs_2, msg=f"failed for {scheduler_func} in save and reload")
```

File: src/transformers/optimization.py (hunk 1 above)
Comment (connor-henderson) [reply src/transformers/optimization.py:53]: Yea visual consistency with the others was the only reason, but that isn't much of one will remove

**Comment by connor-henderson** [comment]
> It may be a good idea to test this feature:

Thank you for filing the issue and sharing this test! I'll leave the decision of whether we should include this test to the maintainers. It may come down to tests being explicitly for functionality as opposed to something like picklability.

**Comment by sgugger** [comment]
Yes it would be nice to have a test such as the one above.

File: tests/optimization/test_optimization.py [hunk 2]
```diff
@@ -17,6 +17,9 @@
 import os
 import tempfile
 import unittest
+from typing import Type
+
+from torch.optim.lr_scheduler import LambdaLR
```
Comment (sgugger) [inline tests/optimization/test_optimization.py:22]: This import should be removed to avoid a error when PyTorch is not installed. It's not really necessary anyway.

File: tests/optimization/test_optimization.py [hunk 3]
```diff
@@ -166,5 +169,20 @@ def test_schedulers(self):
...
+    """See https://github.com/huggingface/transformers/issues/21689"""
+
+    def __init__(self, fn):
+        self.fn = fn
+
+    def __call__(self, *args, **kwargs):
+        return self.fn(*args, **kwargs)
+
+    @classmethod
+    def wrap_scheduler(self, scheduler: Type[LambdaLR]):
```
Comment (sgugger) [inline tests/optimization/test_optimization.py:187]: Remove the type annotation here to remove the dep on PyTorch.

**Review by sgugger — COMMENTED** [review]
Thanks for iterating. YOu will probably need a rebase on main to fix the quality issue, and there is a problem of dependency on torch highlighted below.

File: tests/optimization/test_optimization.py [hunk 4]
```diff
@@ -166,5 +166,21 @@ def test_schedulers(self):
             )
 
             scheduler = scheduler_func(self.optimizer, **kwargs)
+            if scheduler_func.__name__ != "get_constant_schedule":
+                LambdaScheduleWrapper.wrap_scheduler(scheduler)  # wrap to test picklability of the schedule
```
Comment (connor-henderson) [inline tests/optimization/test_optimization.py:170]: Thank you for the comments and sorry for the delay

Just wanted to call out here that I don't love this conditional, but it's the alternative to doing the unneeded `_get_constant_schedule` partial to get the test to pass. If you prefer a different approach just lmk

**Comment by connor-henderson** [comment]
Thank you for the comments and sorry for the delay

**Comment by sgugger** [comment]
Thanks again for your contribution!
